File: small_problems/medium_1/word_to_digit.rb
=begin
PEDAC
  start 157
  finish 2:15
  time 18min
Problem
  rules
    accept a string input containing words and string versions of numbers 1-9

    convert the numbers into their numeric format in the string
    return a new string
  clarifying questions
    should we assume only numbers one-nine will be used? yes
    should we assume consecutive string-numbers should be a single digit? assume yes
Example: see above
  (e.g "one zero minutes" == "1 0 minutes", not "10 minutes")
Data:
  create hash with word
Algorithm
  create hash with keys one-nine, and values 1-9
  separate periods from each word with a space delimiter
  turn string to an array and for each word
    return the hash look up if the word is included in hash keys
    otherwise return the word
  join result with spaces
  remove spaces before periods
=end

NUMBER_TO_STRING = %w(zero one two three four five six seven eight nine).zip('0'..'9').to_a.to_h

# def word_to_digit(string)
#   string = string.gsub(".", " .")
#   string.split.map do |word|
#     if NUMBER_TO_STRING[word].nil?
#       word
#     else
#       NUMBER_TO_STRING[word]
#     end
#   end.join(" ").gsub(" .", ".")
# end

# Refactored
def word_to_digit(string)
  result = string
  NUMBER_TO_STRING.each do |str_num, num|
    result.gsub!(str_num, num)
  end
  result
end

p word_to_digit('Please call me at five five five one two three four. Thanks.') == 'Please call me at 5 5 5 1 2 3 4. Thanks.'

File: js_small_problems/medium_1/word_to_digit.js
/*
start 12:50
time 1:28
takeaways
  stop over optimizing
  start small and work your way through
  ALGORITHM AND EXAMPLES
  TEST EARLY AND OFTEN

Problem
  input: string containing numbers represented as words
    e.g. 0 == 'zero'
  output: string with 'number words' converted to digits
  rules:
    accept a string containing number words
    convert the number words to digits
    retain other characters unmodified
    if a number word has punctuation, replace the number word but keep punctuation
      e.g. example with five and four. -> example with 5 and 4.
  questions:
    - assume all other characters should not be modified?
    - are there other chars that could be considered part of a digit? (e.g. `-` sign)
    - in the example we maintain spaces, should that always be the case? or should we ever combine
      the digits?
    - should all chars attached to a number word be kept?
    - will each number word be separated by a space? can they ever be contiguous
Examples
  expected inputs:
    one or more number words
    zero number words?
    number with punctuation
  malformed inputs:
    nonstring input?
    another object containing the string?
  missing
  extra
  other edge cases
Data Structure
  array of words for transformation
  const array mapping word chars to corresponding index positions
  regex to match only the word if there's punctuation
Algorithm
  initialize const of numberWords as array
  split string into words
  transform words:
    get wordOnlyChars (to ignore puncutation)
    if wordOnlyChars is a number word
      get index of word
      return replacement
    if not, return original word
  return joined string with spaces
*/

function wordToDigit(string) {
  const numberWords = ['zero', 'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine']
  let words = string.split(' ');

  let updatedString = words.map((orgWord) => {
    let regex = /[a-z]+/i;
    let wordCharsOnly = orgWord.toLowerCase().match(regex)[0];

    if (numberWords.includes(wordCharsOnly)) {
      return orgWord.replace(regex, numberWords.indexOf(wordCharsOnly));
    } else {
      return orgWord;
    }
  });

  return updatedString.join(' ');
}

console.log(wordToDigit('Please call me at five five five one two three four. Thanks.'));
// // "Please call me at 5 5 5 1 2 3 4. Thanks."

console.log(wordToDigit('Please call me at my number. Thanks.'));

console.log(wordToDigit('Please call me at my number at One.'));
